geomesa-convert/geomesa-convert-shp/src/main/scala/org/locationtech/geomesa/convert/shp/ShapefileConverter.scala

```diff
@@ -0,0 +1,130 @@
+
+  override def createEvaluationContext(globalParams: Map[String, Any],
+                                       caches: Map[String, EnrichmentCache],
+                                       counter: Counter): EvaluationContext = {
+    // inject placeholders for shapefile attributes into the evaluation context
+    // used for accessing shapefile properties by name in ShapefileFunctionFactory
+    val shpParams = Map(InputSchemaKey -> Array.empty[String], InputValuesKey -> Array.empty[Any])
+    super.createEvaluationContext(globalParams ++ shpParams, caches, counter)
+  }
+
+  override protected def read(is: InputStream, ec: EvaluationContext): CloseableIterator[Array[Any]] = {
+    CloseWithLogging(is) // we don't use the input stream, just close it
+
+    val path = ec.getInputFilePath.getOrElse {
+      throw new IllegalArgumentException(s"Shapefile converter requires '${EvaluationContext.InputFilePathKey}' " +
+          "to be set in the evaluation context")
+    }
+    val ds = ShapefileConverter.getDataStore(path)
+    val schema = ds.getSchema()
+    val array = Array.ofDim[Any](schema.getAttributeCount + 1)
+    val names = Array.tabulate(schema.getAttributeCount)(i => schema.getDescriptor(i).getLocalName)
+
+    val i = ec.indexOf(InputSchemaKey)
+    val j = ec.indexOf(InputValuesKey)
+
+    if (i == -1 || j == -1) {
+      logger.warn("Input schema not found in evaluation context, shapefile functions " +
+          s"${TextTools.wordList(new ShapefileFunctionFactory().functions.map(_.names.head))} will not be available")
+    } else {
+      ec.set(i, names)
+      ec.set(j, array)
+    }
+
+    new ShapefileIterator(ds, array, ec.counter)
+  }
+}
+
+object ShapefileConverter {
+
+  private val factorySet = new AtomicBoolean(false)
+
+  /**
+    * Creates a URL, needed for the shapefile data store
+    *
+    * @param path input path
+    * @return
+    */
+  def getDataStore(path: String): ShapefileDataStore = {
+    val url = if (PathUtils.isRemote(path)) {
+      // we need to add the hadoop url factories to the JVM to support hdfs, S3, or wasb
+      // we only want to call this once per jvm or it will throw an error
+      if (factorySet.compareAndSet(false, true)) {
+        URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory)
+      }
+      new URL(path)
+    } else {
+      new File(path).toURI.toURL
+    }
+    val params = Collections.singletonMap(ShapefileDataStoreFactory.URLP.key, url)
+    val ds = DataStoreFinder.getDataStore(params).asInstanceOf[ShapefileDataStore]
+    if (ds == null) {
+      throw new IllegalArgumentException(s"Could not read shapefile using path '$path'")
+    }
+    ds
+  }
+
+  /**
+    * Iterator for reading a shapefile
+    *
+    * @param ds data store to read from
+    * @param array attributes array
+    * @param counter line counter
+    */
+  class ShapefileIterator(ds: ShapefileDataStore, array: Array[Any], counter: Counter)
```

Comment:
This class doesn't use the fact that this is a ShapefileDataStore.  Do we happen to have a similar DataStore wrapper elsewhere?

Reply:
i don't think so... it does use the fact that it's a FileDataStore (`ds.getFeatureReader`). it's copying the attributes into an array with a +1 offset, which we don't do anywhere else that i can think of